File: sh_erp_cost_dept/task_management/report/task_intelligence_dashboard/task_intelligence_dashboard.py
# Copyright (c) 2024, Amir Mansi and contributors
# For license information, please see license.txt

"""
Task Intelligence Dashboard
============================
Enterprise-grade Script Report for Tasks.

Features:
  • Colored priority & status badges
  • Auto-calculated Duration (Expected End - Expected Start)
  • Assigned-To Users collected from ToDo records
  • Overdue indicator
  • Aging (days open)
  • KPI summary cards (Open / Working / Overdue / Completed)
  • Optimised SQL with a single aggregate query
"""

import frappe
from frappe import _
from frappe.utils import flt, getdate, nowdate

# ---------------------------------------------------------------------------
# Main entry point
# ---------------------------------------------------------------------------


def execute(filters=None):
	filters = frappe._dict(filters or {})
	columns = _get_columns()
	data = _get_data(filters)
	report_summary = _get_report_summary(data)
	return columns, data, None, None, report_summary


# ---------------------------------------------------------------------------
# Column definitions
# ---------------------------------------------------------------------------


def _get_columns():
	return [
		{
			"fieldname": "task_name",
			"label": _("Task ID"),
			"fieldtype": "Link",
			"options": "Task",
			"width": 130,
		},
		{
			"fieldname": "task_title",
			"label": _("Task Title"),
			"fieldtype": "Data",
			"width": 220,
		},
		{
			"fieldname": "priority",
			"label": _("Priority"),
			"fieldtype": "Data",
			"width": 90,
		},
		{
			"fieldname": "exp_start_date",
			"label": _("Expected Start"),
			"fieldtype": "Date",
			"width": 120,
		},
		{
			"fieldname": "exp_end_date",
			"label": _("Expected End"),
			"fieldtype": "Date",
			"width": 120,
		},
		{
			"fieldname": "duration",
			"label": _("Duration (Days)"),
			"fieldtype": "Int",
			"width": 110,
		},
		{
			"fieldname": "status",
			"label": _("Status"),
			"fieldtype": "Data",
			"width": 110,
		},
		{
			"fieldname": "assigned_to_users",
			"label": _("Assigned To"),
			"fieldtype": "Data",
			"width": 200,
		},
		{
			"fieldname": "completed_on",
			"label": _("Completed On"),
			"fieldtype": "Date",
			"width": 120,
		},
		{
			"fieldname": "completed_by",
			"label": _("Completed By"),
			"fieldtype": "Data",
			"width": 150,
		},
		{
			"fieldname": "age_days",
			"label": _("Aging (Days)"),
			"fieldtype": "Int",
			"width": 100,
		},
		{
			"fieldname": "task_creator",
			"label": _("Created By"),
			"fieldtype": "Data",
			"width": 150,
		},
		{
			"fieldname": "is_overdue",
			"label": _("Overdue"),
			"fieldtype": "Data",
			"width": 80,
		},
	]


# ---------------------------------------------------------------------------
# Data query
# ---------------------------------------------------------------------------


def _get_data(filters):
	conditions, params = _build_conditions(filters)
	today = nowdate()

	# A single SQL query that fetches task data and aggregates assigned users
	# from the ToDo table in one pass (LEFT JOIN + GROUP_CONCAT).
	sql = f"""
		SELECT
			t.name                    AS task_name,
			t.subject                 AS task_title,
			t.priority,
			t.exp_start_date,
			t.exp_end_date,
			CASE
				WHEN t.exp_end_date IS NOT NULL AND t.exp_start_date IS NOT NULL
				THEN DATEDIFF(t.exp_end_date, t.exp_start_date)
				ELSE NULL
			END                       AS duration,
			t.status,
			COALESCE(t.custom_assigned_to_users,
				GROUP_CONCAT(DISTINCT td.allocated_to
					ORDER BY td.allocated_to SEPARATOR ', ')
			)                         AS assigned_to_users,
			t.completed_on,
			t.completed_by            AS completed_by,
			DATEDIFF(%(today)s, DATE(t.creation))  AS age_days,
			t.owner                   AS task_creator,
			CASE
				WHEN t.exp_end_date < %(today)s
				  AND t.status NOT IN ('Completed', 'Cancelled', 'Template')
				THEN 1 ELSE 0
			END                       AS is_overdue
		FROM
			`tabTask` t
			LEFT JOIN `tabToDo` td
				ON  td.reference_type = 'Task'
				AND td.reference_name = t.name
				AND td.status = 'Open'
		WHERE
			t.docstatus < 2
			{conditions}
		GROUP BY
			t.name
		ORDER BY
			FIELD(t.priority, 'Urgent', 'High', 'Medium', 'Low'),
			t.exp_end_date ASC
	"""

	params["today"] = today
	rows = frappe.db.sql(sql, params, as_dict=True)
	return rows


def _build_conditions(filters):
	conditions = ""
	params = {}

	if filters.get("from_date"):
		conditions += " AND t.creation >= %(from_date)s"
		params["from_date"] = filters.from_date

	if filters.get("to_date"):
		conditions += " AND t.creation <= %(to_date)s"
		params["to_date"] = filters.to_date

	if filters.get("status"):
		conditions += " AND t.status = %(status)s"
		params["status"] = filters.status

	if filters.get("priority"):
		conditions += " AND t.priority = %(priority)s"
		params["priority"] = filters.priority

	if filters.get("assigned_to"):
		conditions += (
			" AND EXISTS ("
			" SELECT 1 FROM `tabToDo` td2"
			" WHERE td2.reference_type = 'Task'"
			" AND td2.reference_name = t.name"
			" AND td2.allocated_to = %(assigned_to)s"
			" AND td2.status = 'Open'"
			")"
		)
		params["assigned_to"] = filters.assigned_to

	if filters.get("project"):
		conditions += " AND t.project = %(project)s"
		params["project"] = filters.project

	return conditions, params


# ---------------------------------------------------------------------------
# KPI summary (displayed as coloured cards above the report)
# ---------------------------------------------------------------------------


def _get_report_summary(data):
	"""
	Returns a list of KPI dicts for Frappe's built-in report_summary card row.
	Each card shows a count for a specific task status.
	"""
	totals = {
		"open": 0,
		"working": 0,
		"overdue": 0,
		"completed": 0,
	}

	for row in data:
		status = (row.get("status") or "").lower()
		is_overdue = row.get("is_overdue", 0)

		if is_overdue:
			totals["overdue"] += 1
		elif status == "open":
			totals["open"] += 1
		elif status == "working":
			totals["working"] += 1
		elif status == "completed":
			totals["completed"] += 1

	return [
		{
			"value": totals["open"],
			"indicator": "grey",
			"label": _("Open Tasks"),
			"datatype": "Int",
		},
		{
			"value": totals["working"],
			"indicator": "blue",
			"label": _("Working Tasks"),
			"datatype": "Int",
		},
		{
			"value": totals["overdue"],
			"indicator": "red",
			"label": _("Overdue Tasks"),
			"datatype": "Int",
		},
		{
			"value": totals["completed"],
			"indicator": "green",
			"label": _("Completed Tasks"),
			"datatype": "Int",
		},
	]

File: sh_erp_cost_dept/task_management/report/task_intelligence_dashboard/task_intelligence_dashboard.js
// Copyright (c) 2024, Amir Mansi and contributors
// For license information, please see license.txt

/**
 * Task Intelligence Dashboard
 * ----------------------------
 * Client-side filter configuration and custom column formatter for the
 * enterprise Task report.  Provides:
 *   • Filter panel (date range, status, priority, assigned-to, project)
 *   • Colored status badges
 *   • Colored priority badges
 *   • Overdue indicator (red flag icon)
 *   • Assigned-to user badges
 */

frappe.query_reports["Task Intelligence Dashboard"] = {
	// ------------------------------------------------------------------ //
	// Filters
	// ------------------------------------------------------------------ //
	filters: [
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
		},
		{
			fieldname: "status",
			label: __("Status"),
			fieldtype: "Select",
			options:
				"\nOpen\nWorking\nPending Review\nOverdue\nTemplate\nCompleted\nCancelled",
		},
		{
			fieldname: "priority",
			label: __("Priority"),
			fieldtype: "Select",
			options: "\nLow\nMedium\nHigh\nUrgent",
		},
		{
			fieldname: "assigned_to",
			label: __("Assigned To"),
			fieldtype: "Link",
			options: "User",
		},
		{
			fieldname: "project",
			label: __("Project"),
			fieldtype: "Link",
			options: "Project",
		},
	],

	// ------------------------------------------------------------------ //
	// Column formatter – renders badges and indicators
	// ------------------------------------------------------------------ //
	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		if (!data) return value;

		// ---- Status badge -------------------------------------------- //
		if (column.fieldname === "status") {
			const STATUS_COLORS = {
				Open: "#6c757d",
				Working: "#007bff",
				"Pending Review": "#fd7e14",
				Overdue: "#dc3545",
				Template: "#6f42c1",
				Completed: "#28a745",
				Cancelled: "#343a40",
			};
			const raw = (data.status || "").trim();
			const color = STATUS_COLORS[raw] || "#6c757d";
			return `<span style="
				background:${color};
				color:#fff;
				padding:3px 10px;
				border-radius:12px;
				font-size:11px;
				font-weight:600;
				white-space:nowrap;
			">${frappe.utils.escape_html(raw)}</span>`;
		}

		// ---- Priority badge ------------------------------------------ //
		if (column.fieldname === "priority") {
			const PRIORITY_COLORS = {
				Low: "#17a2b8",
				Medium: "#ffc107",
				High: "#fd7e14",
				Urgent: "#dc3545",
			};
			const raw = (data.priority || "").trim();
			const color = PRIORITY_COLORS[raw] || "#6c757d";
			return `<span style="
				background:${color};
				color:${raw === "Medium" ? "#000" : "#fff"};
				padding:3px 10px;
				border-radius:12px;
				font-size:11px;
				font-weight:600;
				white-space:nowrap;
			">${frappe.utils.escape_html(raw)}</span>`;
		}

		// ---- Overdue indicator --------------------------------------- //
		if (column.fieldname === "is_overdue") {
			if (data.is_overdue) {
				return `<span style="color:#dc3545;font-size:16px;" title="${__("Overdue")}">&#9888;</span>`;
			}
			return `<span style="color:#28a745;font-size:14px;" title="${__("On Time")}">&#10003;</span>`;
		}

		// ---- Assigned-to user badges --------------------------------- //
		if (column.fieldname === "assigned_to_users") {
			const raw = (data.assigned_to_users || "").trim();
			if (!raw) return `<span style="color:#aaa;">${__("Unassigned")}</span>`;

			const users = raw.split(",").map((u) => u.trim()).filter(Boolean);
			const badges = users
				.map(
					(u) =>
						`<span style="
							background:#e9ecef;
							color:#495057;
							padding:2px 8px;
							border-radius:10px;
							font-size:11px;
							margin-right:3px;
							display:inline-block;
						">${frappe.utils.escape_html(u)}</span>`
				)
				.join("");
			return badges;
		}

		// ---- Task Title with priority colour ----------------------- //
		if (column.fieldname === "task_title") {
			const PRIORITY_TEXT = {
				Low: "#17a2b8",
				Medium: "#856404",
				High: "#fd7e14",
				Urgent: "#dc3545",
			};
			const color = PRIORITY_TEXT[data.priority] || "inherit";
			return `<span style="color:${color};font-weight:600;">${frappe.utils.escape_html(
				data.task_title || ""
			)}</span>`;
		}

		return value;
	},

	// ------------------------------------------------------------------ //
	// Custom table style injected once when report loads
	// ------------------------------------------------------------------ //
	onload: function (report) {
		_inject_dashboard_styles();
	},
};

// ---------------------------------------------------------------------------
// CSS injection for the report table
// ---------------------------------------------------------------------------

function _inject_dashboard_styles() {
	if (document.getElementById("tid-report-styles")) return;

	const style = document.createElement("style");
	style.id = "tid-report-styles";
	style.textContent = `
		/* Task Intelligence Dashboard – custom table styles */
		.datatable .dt-row:hover .dt-cell {
			background: #f0f7ff !important;
		}
		.datatable .dt-header .dt-cell__content {
			background: #343a40 !important;
			color: #fff !important;
			font-weight: 700;
			font-size: 12px;
			padding: 8px 6px;
		}
		/* Sticky header */
		.datatable thead {
			position: sticky;
			top: 0;
			z-index: 10;
		}
	`;
	document.head.appendChild(style);
}
